library/core/src/num/uint_macros.rs

```diff
@@ -995,6 +1018,32 @@ macro_rules! uint_impl {
             intrinsics::saturating_add(self, rhs)
         }
 
+        /// Saturating addition with a signed integer. Computes `self + rhs`,
+        /// saturating at the numeric bounds instead of overflowing.
+        ///
+        /// # Examples
+        ///
+        /// Basic usage:
+        ///
+        /// ```
+        /// # #![feature(mixed_integer_ops)]
+        #[doc = concat!("assert_eq!(1", stringify!($SelfT), ".saturating_add_signed(2), 3);")]
+        #[doc = concat!("assert_eq!(1", stringify!($SelfT), ".saturating_add_signed(-2), 0);")]
+        #[doc = concat!("assert_eq!((", stringify!($SelfT), "::MAX - 2).saturating_add_signed(4), ", stringify!($SelfT), "::MAX);")]
+        /// ```
+        #[unstable(feature = "mixed_integer_ops", issue = "87840")]
+        #[rustc_const_unstable(feature = "mixed_integer_ops", issue = "87840")]
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline]
+        pub const fn saturating_add_signed(self, rhs: $SignedT) -> Self {
+            if rhs >= 0 {
+                self.saturating_add(rhs as Self)
+            } else {
+                self.saturating_sub(rhs.unsigned_abs())
+            }
```

Comment:
```suggestion
            let (res, overflowed) = self.overflowing_add(rhs as Self);
            if overflowed == (rhs < 0) {
                res
            } else if overflowed {
                0
            } else {
                Self::MAX
            }
```

<details><summary>This version generates branchless code on x86_64</summary>

```asm
playground::saturating_add_signed: # @playground::saturating_add_signed
# %bb.0:
	test	rsi, rsi
	js	.LBB0_1
# %bb.2:
	add	rdi, rsi
	mov	rax, -1
	cmovae	rax, rdi
	ret

.LBB0_1:
	neg	rsi
	xor	eax, eax
	sub	rdi, rsi
	cmovae	rax, rdi
	ret
                                        # -- End function

playground::saturating_add_signed_2: # @playground::saturating_add_signed_2
# %bb.0:
	xor	eax, eax
	add	rdi, rsi
	setb	al
	test	rsi, rsi
	setns	cl
	xor	cl, al
	neg	rax
	test	cl, cl
	cmovne	rax, rdi
	ret
                                        # -- End function
```

</details>

<details><summary>and it is also considerably faster (on x86_64; not sure about aarch64).</summary>

```rust
#![feature(test)]
extern crate test;

use test::bench::{Bencher, black_box};
use rand::{Rng, thread_rng};
use lazy_static::lazy_static;
use std::mem::transmute;

pub fn saturating_add_signed(a: u64, b: i64) -> u64 {
    if b >= 0 {
        a.saturating_add(b as _)
    } else {
        a.saturating_sub(b.unsigned_abs())
    }
}

pub fn saturating_add_signed_2(a: u64, b: i64) -> u64 {
    let (res, overflowed) = a.overflowing_add(b as _);
    if overflowed == (b < 0) {
        res
    } else if overflowed /* && b >= 0 */ {
        u64::MAX
    } else {
        0
    }
}

lazy_static! {
    static ref TEST_NUMBERS: [u128; 4096] = (|| {
        let mut v = [0; 4096];
        thread_rng().fill(&mut v[..]);
        v
    })();
}

#[bench]
fn bench_no_op(bencher: &mut Bencher) {
    bencher.iter(|| {
        for tv in &*TEST_NUMBERS {
            let (a, b): (u64, u64) = unsafe { transmute(*tv) };
            black_box(a);
            black_box(b);
        }
    });
}

#[bench]
fn bench_this_pr(bencher: &mut Bencher) {
    bencher.iter(|| {
        for tv in &*TEST_NUMBERS {
            let (a, b): (u64, i64) = unsafe { transmute(*tv) };
            black_box(saturating_add_signed(black_box(a), black_box(b)));
        }
    });
}

#[bench]
fn bench_suggestion(bencher: &mut Bencher) {
    bencher.iter(|| {
        for tv in &*TEST_NUMBERS {
            let (a, b): (u64, i64) = unsafe { transmute(*tv) };
            black_box(saturating_add_signed_2(black_box(a), black_box(b)));
        }
    });
}
```

</details>

```
test bench_no_op      ... bench:       2,504 ns/iter (+/- 68)
test bench_suggestion ... bench:       5,754 ns/iter (+/- 952)
test bench_this_pr    ... bench:      10,990 ns/iter (+/- 2,130)
```

Reply:
Nice optimization !

I think you swapped the last two branches, though.